File: src/PS2Challenge.Backend/Services/VoteService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using PS2Challenge.Backend.Data;
using PS2Challenge.Backend.Models;
using System.Security.Cryptography;

namespace PS2Challenge.Backend.Services;

public class VoteService
{
    private readonly IServiceScopeFactory _scopeFactory;

    public VoteService(IServiceScopeFactory scopeFactory)
    {
        _scopeFactory = scopeFactory;
    }

    /// <summary>
    /// Archives current votes to vote history with the next available round number
    /// </summary>
    /// <param name="notes">Optional notes about the voting round</param>
    /// <param name="manualPositions">Optional manual position assignments (GameId -> Position)</param>
    /// <returns>Tuple containing the round number and count of archived votes</returns>
    public virtual async Task<(int roundNumber, int archivedCount)> ArchiveCurrentVotesAsync(string? notes = null, Dictionary<int, int>? manualPositions = null)
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        // Get all current votes
        var currentVotes = await db.CurrentVotes.ToListAsync();
        if (!currentVotes.Any())
        {
            throw new InvalidOperationException("No current votes to archive");
        }

        // Get the next round number (max round + 1, or 1 if no history exists)
        var maxRound = await db.VoteHistory.AnyAsync()
            ? await db.VoteHistory.MaxAsync(vh => vh.VoteRound)
            : 0;

        var nextRound = maxRound + 1;

        // Sort current votes by vote count (descending) to determine automatic positions
        var sortedVotes = currentVotes.OrderByDescending(cv => cv.VoteCount).ToList();

        // Create vote history entries
        var historyEntries = new List<VoteHistory>();
        for (int i = 0; i < sortedVotes.Count; i++)
        {
            var cv = sortedVotes[i];
            var position = DeterminePosition(sortedVotes, i, cv, manualPositions);

            historyEntries.Add(new VoteHistory
            {
                GameId = cv.GameId,
                VoteRound = nextRound,
                VoteCount = cv.VoteCount,
                Position = position,
                Notes = string.IsNullOrWhiteSpace(notes) ? null : notes.Trim()
            });
        }

        // Add to history
        await db.VoteHistory.AddRangeAsync(historyEntries);

        // Clear current votes
        db.CurrentVotes.RemoveRange(currentVotes);

        await db.SaveChangesAsync();

        return (nextRound, historyEntries.Count);
    }

    /// <summary>
    /// Gets all current votes with game titles
    /// </summary>
    public virtual async Task<List<CurrentVoteDto>> GetCurrentVotesAsync()
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        var current = await db.CurrentVotes.AsNoTracking().ToListAsync();
        if (!current.Any())
        {
            return new List<CurrentVoteDto>();
        }

        var gameIds = current.Select(c => c.GameId).Distinct().ToList();
        var games = await db.Games.AsNoTracking().Where(g => gameIds.Contains(g.Id)).ToListAsync();
        var idToTitle = games.ToDictionary(g => g.Id, g => g.Title);

        return current
            .Select(c => new CurrentVoteDto
            {
                GameId = c.GameId,
                GameTitle = idToTitle.TryGetValue(c.GameId, out var t) ? t : string.Empty,
                VoteCount = c.VoteCount,
                GameNumber = c.GameNumber
            })
            .OrderBy(x => x.GameNumber)
            .ThenByDescending(x => x.VoteCount)
            .ThenBy(x => x.GameTitle)
            .ToList();
    }

    /// <summary>
    /// Sets current votes (upsert operation)
    /// </summary>
    public virtual async Task<(int inserted, int updated)> SetCurrentVotesAsync(List<CurrentVoteDto> votes)
    {
        if (votes == null || !votes.Any())
        {
            throw new ArgumentException("No votes provided", nameof(votes));
        }

        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        // Collect all titles (normalize)
        var titles = votes
            .Select(v => v.GameTitle?.Trim() ?? string.Empty)
            .Where(t => !string.IsNullOrEmpty(t))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();

        if (!titles.Any())
        {
            throw new ArgumentException("No game titles provided");
        }

        // Load all games into memory and build a case-insensitive map
        var allGames = await db.Games.AsNoTracking().ToListAsync();
        var titleToId = allGames
            .Where(g => !string.IsNullOrWhiteSpace(g.Title))
            .ToDictionary(g => g.Title.Trim(), g => g.Id, StringComparer.OrdinalIgnoreCase);

        // Determine missing titles
        var missing = titles.Where(t => !titleToId.ContainsKey(t)).ToList();
        if (missing.Any())
        {
            throw new InvalidOperationException($"Some game titles were not found: {string.Join(", ", missing)}");
        }

        var gameIds = titleToId.Values.Distinct().ToList();

        // Load existing current votes for the relevant game ids
        var existingDict = await db.CurrentVotes
            .Where(cv => gameIds.Contains(cv.GameId))
            .ToDictionaryAsync(cv => cv.GameId, cv => cv);

        var insertedCount = 0;
        var updatedCount = 0;
        var toInsert = new List<CurrentVote>();

        foreach (var v in votes)
        {
            if (!TryResolveGameId(v, titleToId, out var gameId))
            {
                continue;
            }

            ApplyCurrentVote(
                v,
                gameId,
                existingDict,
                toInsert,
                ref insertedCount,
                ref updatedCount);
        }

        await db.CurrentVotes.AddRangeAsync(toInsert);

        await db.SaveChangesAsync();

        return (insertedCount, updatedCount);
    }

    /// <summary>
    /// Removes a current vote by game title
    /// </summary>
    public virtual async Task<bool> RemoveCurrentVoteAsync(string gameTitle)
    {
        if (string.IsNullOrWhiteSpace(gameTitle))
        {
            throw new ArgumentException("Game title is required", nameof(gameTitle));
        }

        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        // Find game by title (case-insensitive)
        var game = await db.Games
            .AsNoTracking()
            .FirstOrDefaultAsync(g => g.Title.ToLower() == gameTitle.Trim().ToLower());

        if (game == null)
        {
            throw new InvalidOperationException($"Game '{gameTitle}' not found");
        }

        // Find and remove the current vote for this game
        var currentVote = await db.CurrentVotes
            .FirstOrDefaultAsync(cv => cv.GameId == game.Id);

        if (currentVote == null)
        {
            return false;
        }

        db.CurrentVotes.Remove(currentVote);
        await db.SaveChangesAsync();

        return true;
    }

    private static int? DeterminePosition(List<CurrentVote> sortedVotes, int index, CurrentVote vote, Dictionary<int, int>? manualPositions)
    {
        if (manualPositions != null && manualPositions.TryGetValue(vote.GameId, out var manualPos))
        {
            return manualPos;
        }

        return index switch
        {
            0 when sortedVotes.Count == 1 || sortedVotes[1].VoteCount != vote.VoteCount => 1,
            1 when sortedVotes[0].VoteCount != vote.VoteCount && (sortedVotes.Count == 2 || sortedVotes[2].VoteCount != vote.VoteCount) => 2,
            2 when sortedVotes[1].VoteCount != vote.VoteCount => 3,
            _ => null
        };
    }

    private static bool IsValidCurrentVoteInput(CurrentVoteDto vote, string title)
    {
        return !string.IsNullOrEmpty(title)
            && vote.VoteCount >= 0
            && vote.GameNumber >= 1
            && vote.GameNumber <= 3;
    }

    private static bool TryResolveGameId(CurrentVoteDto vote, Dictionary<string, int> titleToId, out int gameId)
    {
        var rawTitle = vote.GameTitle?.Trim() ?? string.Empty;
        gameId = 0;

        return IsValidCurrentVoteInput(vote, rawTitle)
            && titleToId.TryGetValue(rawTitle, out gameId);
    }

    private static void ApplyCurrentVote(
        CurrentVoteDto vote,
        int gameId,
        Dictionary<int, CurrentVote> existingDict,
        List<CurrentVote> toInsert,
        ref int insertedCount,
        ref int updatedCount)
    {
        if (existingDict.TryGetValue(gameId, out var existingRow))
        {
            if (existingRow.VoteCount == vote.VoteCount && existingRow.GameNumber == vote.GameNumber)
            {
                return;
            }

            existingRow.VoteCount = vote.VoteCount;
            existingRow.GameNumber = vote.GameNumber;
            updatedCount++;
            return;
        }

        toInsert.Add(new CurrentVote
        {
            GameId = gameId,
            VoteCount = vote.VoteCount,
            GameNumber = vote.GameNumber
        });
        insertedCount++;
    }

    /// <summary>
    /// Gets vote history grouped by round
    /// </summary>
    public virtual async Task<List<VoteRoundDto>> GetVoteHistoryAsync()
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        // Get all vote history with position
        var allHistory = await db.VoteHistory
            .AsNoTracking()
            .ToListAsync();

        var rounds = allHistory.Select(h => h.VoteRound).Distinct().OrderByDescending(r => r).ToList();

        // Preload game titles for used game ids
        var allGameIds = allHistory.Select(h => h.GameId).Distinct().ToList();
        var games = await db.Games.AsNoTracking().Where(g => allGameIds.Contains(g.Id)).ToListAsync();
        var idToTitle = games.ToDictionary(g => g.Id, g => g.Title);

        var result = new List<VoteRoundDto>();

        foreach (var round in rounds)
        {
            var perRound = allHistory.Where(h => h.VoteRound == round).ToList();

            // Try to use position first, fallback to vote count ordering
            var top = perRound.FirstOrDefault(h => h.Position == 1);
            var second = perRound.FirstOrDefault(h => h.Position == 2);
            var last = perRound.FirstOrDefault(h => h.Position == 3);

            // Fallback: if positions aren't set, use vote count
            if (top == null || second == null || last == null)
            {
                var ordered = perRound.OrderByDescending(h => h.VoteCount).ToList();
                top ??= ordered.ElementAtOrDefault(0);
                second ??= ordered.ElementAtOrDefault(1);
                last ??= ordered.ElementAtOrDefault(2);
            }

            // Pick notes for the round
            var notesForRound = perRound
                .Where(h => !string.IsNullOrEmpty(h.Notes))
                .Select(h => h.Notes)
                .FirstOrDefault();

            string GetTitle(int gameId) => idToTitle.TryGetValue(gameId, out var t) ? t : string.Empty;

            result.Add(new VoteRoundDto
            {
                VoteRound = round,
                TopGameTitle = top != null ? GetTitle(top.GameId) : string.Empty,
                TopVotes = top?.VoteCount ?? 0,
                TopPosition = top?.Position,
                SecondGameTitle = second != null ? GetTitle(second.GameId) : string.Empty,
                SecondVotes = second?.VoteCount ?? 0,
                SecondPosition = second?.Position,
                LastGameTitle = last != null ? GetTitle(last.GameId) : string.Empty,
                LastVotes = last?.VoteCount ?? 0,
                LastPosition = last?.Position,
                Notes = notesForRound ?? string.Empty
            });
        }

        return result;
    }

    /// <summary>
    /// Gets all game titles for autocomplete/suggestions
    /// </summary>
    public virtual async Task<List<string>> GetAllGameTitlesAsync()
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        return await db.Games
            .AsNoTracking()
            .Select(g => g.Title)
            .Where(t => !string.IsNullOrWhiteSpace(t))
            .Distinct()
            .ToListAsync();
    }

    /// <summary>
    /// Fills current votes with random eligible games
    /// </summary>
    /// <param name="count">Number of games to add (will be limited by available slots)</param>
    /// <returns>Number of games added</returns>
    public virtual async Task<int> FillCurrentVotesWithRandomGamesAsync(int count)
    {
        if (count <= 0)
        {
            throw new ArgumentException("Count must be greater than 0", nameof(count));
        }

        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<Ps2ChallengeDbContext>();

        // Get current vote count
        var currentVotesCount = await db.CurrentVotes.CountAsync();

        // Calculate how many games we can add (max 3 total)
        var slotsAvailable = 3 - currentVotesCount;
        if (slotsAvailable <= 0)
        {
            throw new InvalidOperationException("Current votes already has 3 games. Archive or remove existing votes first.");
        }

        var gamesToAdd = Math.Min(count, slotsAvailable);

        // Get owned game IDs
        var ownedGameIds = await db.GamesOwned
            .Select(go => go.GameId)
            .ToListAsync();

        // Get excluded game IDs
        var excludedGameIds = await db.ExcludedGames
            .Select(eg => eg.GameId)
            .ToListAsync();

        // Get games that have been started (have progress records)
        var startedGameIds = await db.GameProgress
            .Select(gp => gp.GameId)
            .ToListAsync();

        // Get game IDs already in current votes
        var currentVoteGameIds = await db.CurrentVotes
            .Select(cv => cv.GameId)
            .ToListAsync();

        // Get eligible game IDs: owned, not excluded, not started, not in current votes
        var eligibleGameIds = ownedGameIds
            .Where(id => !excludedGameIds.Contains(id))
            .Where(id => !startedGameIds.Contains(id))
            .Where(id => !currentVoteGameIds.Contains(id))
            .ToList();

        if (!eligibleGameIds.Any())
        {
            throw new InvalidOperationException("No eligible games found. Games must be owned, not excluded, and not started.");
        }

        if (eligibleGameIds.Count < gamesToAdd)
        {
            throw new InvalidOperationException($"Only {eligibleGameIds.Count} eligible game(s) available, but {gamesToAdd} requested.");
        }

        // Randomly select games
        var shuffledGameIds = eligibleGameIds.ToList();
        for (var i = shuffledGameIds.Count - 1; i > 0; i--)
        {
            var j = RandomNumberGenerator.GetInt32(i + 1);
            (shuffledGameIds[i], shuffledGameIds[j]) = (shuffledGameIds[j], shuffledGameIds[i]);
        }

        var selectedGameIds = shuffledGameIds
            .Take(gamesToAdd)
            .ToList();

        // Get used game numbers
        var usedGameNumbers = await db.CurrentVotes
            .Select(cv => cv.GameNumber)
            .ToListAsync();

        // Find available game numbers (1-3)
        var availableGameNumbers = Enumerable.Range(1, 3)
            .Where(n => !usedGameNumbers.Contains(n))
            .ToList();

        // Create new current vote entries
        var newVotes = new List<CurrentVote>();
        for (int i = 0; i < selectedGameIds.Count; i++)
        {
            newVotes.Add(new CurrentVote
            {
                GameId = selectedGameIds[i],
                VoteCount = 0,
                GameNumber = availableGameNumbers[i]
            });
        }

        await db.CurrentVotes.AddRangeAsync(newVotes);
        await db.SaveChangesAsync();

        return newVotes.Count;
    }
}

File: apps/server/src/services/voteService.ts
import crypto from "node:crypto";
import type { CurrentVoteDto, UploadRoundDto, VoteRoundDto } from "@ps2-challenge/shared";
import { sql, type Kysely } from "kysely";
import type pg from "pg";
import { createKyselyFromPool, type Database } from "../db/kysely.js";

type CurrentVoteRow = {
  game_id: number;
  vote_count: number;
  game_number: number;
};

type VoteHistoryRow = {
  game_id: number;
  game_title: string;
  vote_round: number;
  vote_count: number;
  position: number | null;
  notes: string | null;
};

export class VoteService {
  private readonly db: Kysely<Database>;

  constructor(pool: pg.Pool, db?: Kysely<Database>) {
    this.db = db ?? createKyselyFromPool(pool);
  }

  async getCurrentVotes(): Promise<CurrentVoteDto[]> {
    const rows = await this.db
      .selectFrom("current_vote as cv")
      .leftJoin("games as g", "g.game_id", "cv.game_id")
      .select([
        "cv.game_id as game_id",
        sql<string>`COALESCE(g.title, '')`.as("game_title"),
        "cv.vote_count as vote_count",
        "cv.game_number as game_number"
      ])
      .orderBy("cv.game_number")
      .orderBy("cv.vote_count", "desc")
      .orderBy("g.title")
      .execute();

    return rows.map(mapCurrentVote);
  }

  async setCurrentVotes(votes: CurrentVoteDto[]): Promise<{ inserted: number; updated: number }> {
    if (!votes.length) {
      throw new Error("No votes provided");
    }

    const titles = [...new Set(votes.map((vote) => vote.gameTitle.trim()).filter(Boolean).map((title) => title.toLocaleLowerCase("en-GB")))];
    if (!titles.length) {
      throw new Error("No game titles provided");
    }

    const games = await this.db.selectFrom("games").select(["game_id", "title"]).execute();
    const titleToId = new Map(games.map((row) => [row.title.trim().toLocaleLowerCase("en-GB"), row.game_id]));
    const missing = titles.filter((title) => !titleToId.has(title));
    if (missing.length) {
      throw new Error(`Some game titles were not found: ${missing.join(", ")}`);
    }

    const resolved = votes
      .map((vote) => ({ vote, gameId: titleToId.get(vote.gameTitle.trim().toLocaleLowerCase("en-GB")) ?? 0 }))
      .filter(({ vote, gameId }) => gameId > 0 && vote.voteCount >= 0 && vote.gameNumber >= 1 && vote.gameNumber <= 3);

    const duplicateGameIds = resolved
      .map((entry) => entry.gameId)
      .filter((gameId, index, all) => all.indexOf(gameId) !== index);
    if (duplicateGameIds.length) {
      const duplicateTitles = resolved
        .filter((entry) => duplicateGameIds.includes(entry.gameId))
        .map((entry) => entry.vote.gameTitle.trim());
      throw new Error(`Duplicate game titles are not allowed: ${[...new Set(duplicateTitles)].join(", ")}`);
    }

    const existing = await this.currentVoteRows(this.db);
    const projectedNumbers = new Map(existing.map((row) => [row.game_id, row.game_number]));
    for (const entry of resolved) {
      projectedNumbers.set(entry.gameId, entry.vote.gameNumber);
    }
    const duplicates = [...projectedNumbers.values()]
      .filter((gameNumber, index, all) => all.indexOf(gameNumber) !== index)
      .sort((left, right) => left - right);
    if (duplicates.length) {
      throw new Error(`Current votes must use unique game numbers. Duplicate positions: ${[...new Set(duplicates)].join(", ")}`);
    }

    let inserted = 0;
    let updated = 0;
    await this.db.transaction().execute(async (transaction) => {
      for (const entry of resolved) {
        const current = existing.find((row) => row.game_id === entry.gameId);
        if (current) {
          if (current.vote_count !== entry.vote.voteCount || current.game_number !== entry.vote.gameNumber) {
            await transaction
              .updateTable("current_vote")
              .set({ vote_count: entry.vote.voteCount, game_number: entry.vote.gameNumber })
              .where("game_id", "=", entry.gameId)
              .execute();
            updated++;
          }
        } else {
          await transaction
            .insertInto("current_vote")
            .values({ game_id: entry.gameId, vote_count: entry.vote.voteCount, game_number: entry.vote.gameNumber })
            .execute();
          inserted++;
        }
      }
    });

    return { inserted, updated };
  }

  async removeCurrentVote(gameTitle: string): Promise<boolean> {
    if (!gameTitle.trim()) {
      throw new Error("Game title is required");
    }
    const game = await this.db.selectFrom("games").select("game_id").where("title", "ilike", gameTitle.trim()).executeTakeFirst();
    if (!game) {
      throw new Error(`Game '${gameTitle}' not found`);
    }
    const result = await this.db.deleteFrom("current_vote").where("game_id", "=", game.game_id).executeTakeFirst();
    return (result.numDeletedRows ?? 0n) > 0n;
  }

  async archiveCurrentVotes(notes?: string | null, manualPositions?: Record<number, number>): Promise<{ roundNumber: number; archivedCount: number }> {
    return this.db
      .transaction()
      .setIsolationLevel("serializable")
      .execute(async (transaction) => {
        const current = await this.currentVoteRows(transaction);
        if (!current.length) {
          throw new Error("No current votes to archive");
        }

        const maxRound = await transaction
          .selectFrom("vote_history")
          .select((eb) => eb.fn.max<number>("vote_round").as("max_round"))
          .executeTakeFirst();
        const roundNumber = (maxRound?.max_round ?? 0) + 1;
        const sorted = [...current].sort((left, right) => right.vote_count - left.vote_count);
        for (let index = 0; index < sorted.length; index++) {
          const vote = sorted[index]!;
          const position = determinePosition(sorted, index, vote, manualPositions);
          await transaction
            .insertInto("vote_history")
            .values({
              game_id: vote.game_id,
              vote_round: roundNumber,
              vote_count: vote.vote_count,
              position,
              notes: notes?.trim() || null
            })
            .execute();
        }
        await transaction.deleteFrom("current_vote").execute();
        return { roundNumber, archivedCount: sorted.length };
      });
  }

  async getVoteHistory(): Promise<VoteRoundDto[]> {
    const rows = await this.db
      .selectFrom("vote_history as vh")
      .leftJoin("games as g", "g.game_id", "vh.game_id")
      .select([
        "vh.game_id as game_id",
        sql<string>`COALESCE(g.title, '')`.as("game_title"),
        "vh.vote_round as vote_round",
        "vh.vote_count as vote_count",
        "vh.position as position",
        "vh.notes as notes"
      ])
      .orderBy("vh.vote_round", "desc")
      .execute();

    const history = rows as VoteHistoryRow[];
    const rounds = [...new Set(history.map((row) => row.vote_round))].sort((left, right) => right - left);
    return rounds.map((round) => {
      const perRound = history.filter((row) => row.vote_round === round);
      const ordered = [...perRound].sort((left, right) => right.vote_count - left.vote_count);
      const top = perRound.find((row) => row.position === 1) ?? ordered[0];
      const second = perRound.find((row) => row.position === 2) ?? ordered[1];
      const last = perRound.find((row) => row.position === 3) ?? ordered[2];
      const notes = perRound.find((row) => row.notes)?.notes ?? "";

      return {
        voteRound: round,
        topGameTitle: top?.game_title ?? "",
        topVotes: top?.vote_count ?? 0,
        topPosition: top?.position ?? undefined,
        secondGameTitle: second?.game_title ?? "",
        secondVotes: second?.vote_count ?? 0,
        secondPosition: second?.position ?? undefined,
        lastGameTitle: last?.game_title ?? "",
        lastVotes: last?.vote_count ?? 0,
        lastPosition: last?.position ?? undefined,
        notes
      };
    });
  }

  async uploadHistory(rounds: UploadRoundDto[]): Promise<{ inserted: number; updated: number }> {
    if (!rounds.length) {
      throw new Error("No rounds provided");
    }

    let inserted = 0;
    let updated = 0;
    await this.db.transaction().execute(async (transaction) => {
      for (const round of rounds) {
        for (const vote of round.votes) {
          const game = await transaction.selectFrom("games").select("game_id").where("title", "ilike", vote.gameTitle.trim()).executeTakeFirst();
          if (!game) {
            throw new Error(`Some game titles were not found: ${vote.gameTitle}`);
          }
          const existing = await transaction
            .selectFrom("vote_history")
            .select("history_id")
            .where("vote_round", "=", round.voteRound)
            .where("game_id", "=", game.game_id)
            .executeTakeFirst();
          if (existing) {
            await transaction
              .updateTable("vote_history")
              .set({ vote_count: vote.count, position: vote.position ?? null, notes: round.notes ?? null })
              .where("vote_round", "=", round.voteRound)
              .where("game_id", "=", game.game_id)
              .execute();
            updated++;
          } else {
            await transaction
              .insertInto("vote_history")
              .values({
                game_id: game.game_id,
                vote_round: round.voteRound,
                vote_count: vote.count,
                position: vote.position ?? null,
                notes: round.notes ?? null
              })
              .execute();
            inserted++;
          }
        }
      }
    });
    return { inserted, updated };
  }

  async updateVoteCountByGameNumber(gameNumber: number, voteCount: number): Promise<CurrentVoteDto | null> {
    const row = await this.db
      .updateTable("current_vote")
      .set({ vote_count: voteCount })
      .where("game_number", "=", gameNumber)
      .returning(["game_id", "vote_count", "game_number"])
      .executeTakeFirst();
    if (!row) {
      return null;
    }

    const game = await this.db.selectFrom("games").select("title").where("game_id", "=", row.game_id).executeTakeFirst();
    return { gameId: row.game_id, gameTitle: game?.title ?? "", voteCount: row.vote_count, gameNumber: row.game_number };
  }

  async fillCurrentVotesWithRandomGameDetails(count: number): Promise<CurrentVoteDto[]> {
    if (count <= 0) {
      throw new Error("Count must be greater than 0");
    }

    const current = await this.getCurrentVotes();
    const slotsAvailable = 3 - current.length;
    if (slotsAvailable <= 0) {
      throw new Error("Current votes already has 3 games. Archive or remove existing votes first.");
    }
    const gamesToAdd = Math.min(count, slotsAvailable);
    const eligible = await this.db
      .selectFrom("games as g")
      .innerJoin("game_owned as go", "go.game_id", "g.game_id")
      .leftJoin("excluded_games as eg", "eg.game_id", "g.game_id")
      .leftJoin("progress as p", "p.game_id", "g.game_id")
      .leftJoin("current_vote as cv", "cv.game_id", "g.game_id")
      .select(["g.game_id as game_id", "g.title as title"])
      .where("eg.game_id", "is", null)
      .where("p.game_id", "is", null)
      .where("cv.game_id", "is", null)
      .execute();
    if (!eligible.length) {
      throw new Error("No eligible games found. Games must be owned, not excluded, and not started.");
    }
    if (eligible.length < gamesToAdd) {
      throw new Error(`Only ${eligible.length} eligible game(s) available, but ${gamesToAdd} requested.`);
    }

    const shuffled = shuffle(eligible);
    const usedNumbers = new Set(current.map((vote) => vote.gameNumber));
    const availableNumbers = [1, 2, 3].filter((number) => !usedNumbers.has(number));
    const added: CurrentVoteDto[] = [];
    for (let index = 0; index < gamesToAdd; index++) {
      const game = shuffled[index]!;
      const gameNumber = availableNumbers[index]!;
      await this.db.insertInto("current_vote").values({ game_id: game.game_id, vote_count: 0, game_number: gameNumber }).execute();
      added.push({ gameId: game.game_id, gameTitle: game.title, voteCount: 0, gameNumber });
    }
    return added;
  }

  private currentVoteRows(db: Kysely<Database>): Promise<CurrentVoteRow[]> {
    return db.selectFrom("current_vote").select(["game_id", "vote_count", "game_number"]).execute();
  }
}

function mapCurrentVote(row: { game_id: number; game_title: string; vote_count: number; game_number: number }): CurrentVoteDto {
  return {
    gameId: row.game_id,
    gameTitle: row.game_title,
    voteCount: row.vote_count,
    gameNumber: row.game_number
  };
}

function determinePosition(
  sortedVotes: { game_id: number; vote_count: number }[],
  index: number,
  vote: { game_id: number; vote_count: number },
  manualPositions?: Record<number, number>
): number | null {
  const manual = manualPositions?.[vote.game_id];
  if (manual !== undefined) {
    return manual;
  }
  if (index === 0 && (sortedVotes.length === 1 || sortedVotes[1]!.vote_count !== vote.vote_count)) {
    return 1;
  }
  if (index === 1 && sortedVotes[0]!.vote_count !== vote.vote_count && (sortedVotes.length === 2 || sortedVotes[2]!.vote_count !== vote.vote_count)) {
    return 2;
  }
  if (index === 2 && sortedVotes[1]!.vote_count !== vote.vote_count) {
    return 3;
  }
  return null;
}

function shuffle<T>(items: T[]): T[] {
  const shuffled = [...items];
  for (let index = shuffled.length - 1; index > 0; index--) {
    const swap = crypto.randomInt(index + 1);
    [shuffled[index], shuffled[swap]] = [shuffled[swap]!, shuffled[index]!];
  }
  return shuffled;
}
